### PR TITLE
(#124, #35, #34) Initial Analytics Implementation

### LINKED ISSUE #34: Story: Click Event on A-Z List
## As a user, when I click on an A-Z link I should be raising an analytics event so that the website owner may gather data on my user behavior.

> **ESTIMATE** 3

### Acceptance Criteria

```gherkin
Scenario: Analytics click event on A-Z list is raised when user clicks on a letter in the list.
  Given "audience" is set to "Patient"
    And "dictionaryTitle" is set to "NCI Dictionary of Cancer Terms"
    And "dictionaryName" is set to "Cancer.gov"
    And "language" is set to "en"
    And "baseHost" is set to "http://localhost:3000"
    And "canonicalHost" is set to "https://www.cancer.gov"
    And "siteName" is set to "National Cancer Institute"
    And "channel" is set to "Publications"
    And "publishedDate" is set to "02/02/2011"
    And NCIDataLayer is being captured 
  When the user views the definition with prettyurlname "metastatic"
    And user selects letter "A" from A-Z list
  Then there should be an NCIDataLayer Other event with the name "GlossaryApp:Other:AZClick" and linkname "TermsDictionarySearchAlphaList"
    And the event has the following data
      | key           | value |
      | letter        | A |
```

### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* Web Page: *Here goes a URL to an existing web page*

### Notes
* When an "Other" event type is triggered in Launch, then the last page load will be part of the object. So language, dictionary name, etc will already be there, so there is no need to add it to this event's data.

---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [ ] Short Task Description - Issue #9999
- [ ] Short Task Description - Issue #9999

### Technical Notes
- More notes, or technical nuances/direction

### LINKED ISSUE #35: Story: Term Page Analytics
## As an owner of the page, I would like to collect analytics to know which term pages are being visited so that I can further analyze user related data.

> **ESTIMATE** 3

### Acceptance Criteria

```gherkin
Scenario: Page Load Analytics fires when a user visits a term page.
  Given "audience" is set to "Patient"
    And "dictionaryTitle" is set to "NCI Dictionary of Cancer Terms"
    And "dictionaryName" is set to "Cancer.gov"
    And "language" is set to "en"
    And "baseHost" is set to "http://localhost:3000"
    And "canonicalHost" is set to "https://www.cancer.gov"
    And "siteName" is set to "National Cancer Institute"
    And "channel" is set to "Publications"
    And "publishedDate" is set to "02/02/2011"
    And NCIDataLayer is being captured 
  When the user views the definition with prettyurlname "metastatic"
  Then there should be an NCIDataLayer PageLoad event with the name "GlossaryApp:Load:Definition"
    And the event has the following page details
      | key           | value |
      | name          | www.cancer.gov/def/metastatic |
      | title         | metastatic |
      | metaTitle     | Definition of metastatic - NCI Dictionary of Cancer Terms - National Cancer Institute |
      | language      | english |
      | type          | nciAppModulePage |
      | audience      | Patient |
      | channel       | Publications |
      | contentGroup  | NCI Dictionary of Cancer Terms |
      | publishedDate | 02/02/2011 |
    And the event has the following additional details for the page
      | key             | value |
      | dictionaryName  | NCI Dictionary of Cancer Terms |
      | term            | metastatic |
      | id              | 44058 |

```
### Rules for Analytics
* page data:
  * name - this is the canonical host + path. (Basically the canonical URL without https)
  * title - This will be what is used in the H1 tag. So for a definition this would be the term name.
  * metaTitle - This should be whatever is going to be set as the `<title>` for the page.
  * language - `english` or `spanish` depending on the language
  * type - this is hardcoded to `nciAppModulePage`
  * audience - Discussed with Joe, we can use what is passed in, following the proper naming of the items
  * channel - this should be passed in to the app, and in a Digital Platform page this value comes from the `dcterms.subject` meta tag.
  * contentGroup - Discussed with Joe we will just use the Dictionary Title for this since it is pretty much the same thing. (and if the app moves in the future we will not forget to change the new site section...)
  * publishedDate - This should be passed in to the app. It comes from  `meta[name="dcterms.issued"]`. It does need to be in the format MM/DD/YYYY - Drupal adds a timestamp and needs to be cleaned up prior to the app.  
* additional details (page.additionalDetails)
   * dictionaryName - this should be the dictionaryName 
   * term - this should be the term name
   * id - this should be the term id

---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [ ] Short Task Description - Issue #9999
- [ ] Short Task Description - Issue #9999

### Technical Notes
- More notes, or technical nuances/direction

### LINKED ISSUE #124: Enabler: Update Analytics Handler to push to NCIDataLayer
# The Analytics handler needs to be updated for the NCI Event-Driven Data Layer (EDDL)

## Background
Here is some background to understand how the new EDDL works, and how react-tracker works to understand the approach

### New Analytics Event-Driven Data Layer
The new way of handling analytics requires that the following code be used for a page load event:
```
window.NCIDataLayer = window.NCIDataLayer || [];
window.NCIDataLayer.push({
  type: 'PageLoad',
  event: '<EVENT_NAME>',
  page: {
    name: "",
    title: "",
    metaTitle: "",
    language: "",
    type: "",
    audience: "",
    channel: "",
    contentGroup: "",
    publishedDate: Date
    additionalDetails: {}
  }
});
```

and the following for click events:
```
window.NCIDataLayer.push({
  type: 'Other',
  event: '<EVENT_NAME>',
  data: {

  }
})
```

One of the MOST IMPORTANT things is that page load events ALWAYS preceed click events. The EDDL keeps track of the page information raised during a page load, and that information is pushed out to the Analytics Tool with the click/other data payload. So if a click event is raised by the app BEFORE the page load it is associated with, then bad things happen...

### How the react-tracker Package Works
The [react-tracking library](https://github.com/nytimes/react-tracking) offers a way for embedding analytics agnostic event tracking in a react app. React-tracker OOB allows you to set various contextual data points in each of your components, such that if a nested component raises a tracking event, those data points are included in the data payload. 

For example say you are displaying a search results page. You can:
* From the view component for the search results, you `track({ pageName: 'Results Page', searchTerm: 'chicken'})`
* From the results listing component you `track({numResults: 322})`
* Then on a specific result component on a click handler of a link you can 
   ```
   tracking.trackEvent({
      action: 'result_link_click',
      position: thePosition,
      title: result.title
   });
   ```  
Then when a user clicks on a result link a tracking event is dispatched with:
```
{
  pageName: 'Results Page',
  searchTerm: 'chicken',
  numResults: 322,
  action: 'result_link_click',
  position: 3,
  title: 'A title'
}
```

We have created an AnalyticsProvider higher-order component that wraps our App "component". This provider wires up a custom dispatch function to the react-tracking library. This custom function is the `analyticsHandler` parameter passed into the initialize function. 

A react-tracker dispatch function takes in a data payload that has no predefined structured. 

## What To Do?
1. Create a default `analyticsHandler` function that does the following to transform a trackEvent payload into a EDDL payload:
    ```
     // Initialize data layer
     window.NCIDataLayer = window.NCIDataLayer || [];
     (payload) => {
        if (!payload.type || !payload.event) {
          console.error("Malformed analytics event");
          console.error(payload);
        }
        switch(data.type) {
          case 'Other': {
            const { 
               type, event,
               // Strip out items that may be sent with every click event because
               // a developer sets up the component hierarchy to make their life
               // easier. Maybe the channel, language and contentGroup are set at
               // the App level. Meta title at the page level. Who knows. We should
               // offer flexibility and remove noise.  This way the data element can
               // be build up through the nested components as well.
               type, event, name, title, metaTitle, language, 
               audience, channel, contentGroup, publishedDate,
               ...data } = payload;
            window.NCIDataLayer.push({
               type,
               event,
               data: data ? data : { }
            });
            break;
          }
          case 'PageLoad': {
            const { 
               type, event, name, title, metaTitle, language,
               audience, channel, contentGroup, publishedDate,
               ...additionalDetails
            } = payload;
            window.NCIDataLayer.push({
               type, event,
               page: {
                 name, title, metaTitle, language,
                 type: 'nciAppModulePage',
                 audience, channel, contentGroup, publishedDate,
                 additionalDetails: additionalDetails ? additionalDetails : { }
               }
            });
            break;
          }
          default: {
            console.error("Malformed analytics event");
            console.error(payload);
          } 
        }
     }
    ``` 
2. Create some cucumber step definitions to test to ensure the correct objects are pushed onto the NCIDataLayer.

> **ESTIMATE** NN

### Resources:
* EDDL WIP Documentation: [NewLaunchAnalytics.pdf](https://github.com/NCIOCPL/glossary-app/files/4518357/NewLaunchAnalytics.pdf)

### Sub-Tasks
- [ ] Create the analytics handler.
- [ ] Create the testing steps

### PR DESCRIPTION
**Pull request by bryanpizzillo**
Closes #124
Closes #35
Closes #34

### CODE REVIEW

**Comment by bryanpizzillo** [comment]
@kate-mashkina not setting you as a reviewer because you made a lot of the files. BUT FYI.

File: src/index.js [hunk 1]
```diff
@@ -46,36 +52,44 @@ const initialize = ({
...
     dictionaryTitle,
+    analyticsChannel,
+    analyticsPublishedDate,
     language,
     languageToggleSelector,
     searchBoxTitle,
     siteName
   };
 
+  const AppBlock = () => {
```
Comment (arcepaul) [inline src/index.js:63]: 🥇

File: src/tracking/eddl-analytics-provider.jsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  analyticsHandler,
+  pageName,
+  pageTitle,
+  pageMetaTitle,
+  pageLanguage,
+  pageAudience,
+  pageChannel,
+  pageContentGroup,
+  pagePublishedDate,
+  ...pageAdditionalDetails
```
Comment (arcepaul) [inline src/tracking/eddl-analytics-provider.jsx:27]: decided to keep it!

Comment (bryanpizzillo) [reply src/tracking/eddl-analytics-provider.jsx:27]: We actually NEED it after I looked.

**Pull request merged**